at_onboarding_cli/lib/src/at_onboarding_service_impl.dart

```diff
@@ -0,0 +1,276 @@
+import 'dart:convert';
+import 'dart:io';
+import 'package:at_client/at_client.dart';
+import 'package:at_commons/at_builders.dart';
+import 'package:at_commons/at_commons.dart';
+import 'package:at_lookup/at_lookup.dart';
+import 'package:at_server_status/at_server_status.dart';
+import 'package:at_onboarding_cli/at_onboarding_cli.dart';
+import 'package:at_utils/at_logger.dart';
+import 'package:crypton/crypton.dart';
+import 'package:encrypt/encrypt.dart';
+import 'package:zxing2/qrcode.dart';
+import 'package:image/image.dart';
+
+///class containing service that can onboard/activate/authenticate @signs
+class AtOnboardingServiceImpl implements AtOnboardingService {
+  String _atSign;
+  bool _isPkamAuthenticated = false;
+  bool _isAtsignOnboarded = false;
+  AtLookupImpl? _atLookup;
+  AtClient? _atClient;
+  AtSignLogger logger = AtSignLogger('Onboarding Cli');
+  AtOnboardingPreference atOnboardingPreference;
+
+  AtOnboardingServiceImpl(this._atSign, this.atOnboardingPreference);
+
+  @override
+  Future<AtClient?> getAtClient() async {
+    if (_atClient == null) {
+      AtClientManager _atClientManager = AtClientManager.getInstance();
+      await _atClientManager.setCurrentAtSign(
+          _atSign, atOnboardingPreference.namespace, atOnboardingPreference);
+      _atLookup = _atClientManager.atClient.getRemoteSecondary()?.atLookUp;
+      return _atClientManager.atClient;
+    }
+    return _atClient;
+  }
+
+  @override
+  Future<bool> onboard() async {
+    //get cram_secret from either from AtOnboardingConfig or decode it from qr code whichever available
+    atOnboardingPreference.cramSecret ??=
+        _getSecretFromQr(atOnboardingPreference.qrCodePath);
+    if (atOnboardingPreference.cramSecret != null) {
+      _atLookup = AtLookupImpl(_atSign, atOnboardingPreference.rootDomain,
+          atOnboardingPreference.rootPort);
+      _isAtsignOnboarded = (await _atLookup
+          ?.authenticate_cram(atOnboardingPreference.cramSecret))!;
+      if (_isAtsignOnboarded == true) {
+        await _activateAtsign();
+        logger.finer('cram authentication successful');
+        return true;
+      }
+      return false;
+    }
+    throw UnAuthenticatedException(
```

Comment:
we can check the condition at the beginning and throw the UnAuthenticatedException